### DISCUSSION TITLE
runInAction - Should I try to group multiple property updates or not?

### DISCUSSION THREAD

**Discussion by codeBelt** [mobxjs/mobx#3495] [Q&A] [answered]
If I have multiple properties that are being updated should I group all the code I can within `runInAction` or only put the properties that are being updated in `runInAction`? The `init` method examples below could be broken up into smaller methods but in this situation would it better to use two or five `runInAction`'s?

My main question is there any difference between the two examples below in regards to trigger renders within React? Does `runInAction` wait for all properties to be updated before triggering or any update to different properties within `runInAction` will cause updates immediately?

----
1. Have two `runInAction`'s and group most of the code between the `await`'s.

```ts
export class InsightsPageStore {
  // ...
  constructor() {
    makeAutoObservable(this, {}, { autoBind: true });
  }

  async init() {
    await this.loadInitialData();

    runInAction(() => {
      if (this.isEmployeeSearchView) {
        // ...
        if (someCondition) {
          this.selectedRoles = [this.roleId];
          this.selectedDepartments = [this.departmentId];
        }
      }

      if (anotherCondition) {
        this.selectedEmployeeId = Router.query.employee_id || null;
      }
    });

    await Promise.all([this.loadEmployeesSidePanelData(), this.loadEmployeeGridData()]);

    runInAction(() => {
      if (this.selectedEmployeeId) {
        const employee = this.loadedData.find((item) => item.employee.id === this.selectedEmployeeId);

        if (employee) {
          this.searchTerm = buildFullName(employee.employee);
          this.displaySearchTerm = buildFullName(employee.employee);
        }
      }

      this.allAvailableTags = this.otherLoadedData.tags;
      this.loadingState = VirtualShaperLoadingState.None;
    });
  }
}
```
----
2. Have five `runInAction`'s and only use where it is needed.

```ts
export class InsightsPageStore {
  // ...
  constructor() {
    makeAutoObservable(this, {}, { autoBind: true });
  }

  async init() {
    await this.loadInitialData();

    if (this.isEmployeeSearchView) {
      // ...
      if (someCondition) {
        runInAction(() => {
          this.selectedRoles = [this.roleId];
          this.selectedDepartments = [this.departmentId];
        });
      }
    }

    if (anotherCondition) {
      runInAction(() => {
        this.selectedEmployeeId = Router.query.employee_id || null;
      });
    }

    await Promise.all([this.loadEmployeesSidePanelData(), this.loadEmployeeGridData()]);

    if (this.selectedEmployeeId) {
      const employee = this.loadedData.find((item) => item.employee.id === this.selectedEmployeeId);

      if (employee) {
        runInAction(() => {
          this.searchTerm = buildFullName(employee.employee);
          this.displaySearchTerm = buildFullName(employee.employee);
        });
      }
    }

    runInAction(() => {
      this.allAvailableTags = this.otherLoadedData.tags;
      this.loadingState = VirtualShaperLoadingState.None;
    });
  }
}
```

**Comment by kubk** [comment, marked as answer]
The more `runInAction` you have the more re-renders you'll get (unless actions are nested, in this case Mobx will wait till the outermost action is finished). Here is an example: https://codesandbox.io/s/confident-matsumoto-z81rlu

Notice that the render count increases twice slower than the amount of observable changes.

In your case actions aren't nested because the task after await is disconnected from the store's action. So a more performat should be the first example.

To avoid nesting completely you may try mobx flow. Generators has some issues with TS, but there is a workaround (I've been using it in production for a while, it works like a charm): https://github.com/mobxjs/mobx/discussions/3195

**Reply by codeBelt** [reply]
@t49tran thanks, Could you provide an codesandbox example with flow? I am having issue trying to get something to work.

Here is a messy example I am working on https://codesandbox.io/s/hungry-sanne-c8dotp?file=/index.js

**Reply by kubk** [reply]
@codeBelt Done. The rules of using are pretty simple:

- change `async` to `*`
- change `await` to `yield`

I believe we should provide a better documentation for this feature.

If you want it to play nicely with TS then you can grab that helper and replace `await` with `yield* result()`
